Orange/classification/rules.py

```diff
@@ -0,0 +1,938 @@
+import operator
+from copy import copy
+from hashlib import sha1
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+from scipy.stats import chisqprob
+import Orange
+from Orange.classification import Learner, Model
+
+__all__ = ["CN2Learner", "CN2UnorderedLearner"]
+
+
+def argmaxrnd(a, random_seed=None):
+    """
+    Find the index of the maximum value for a given 1-D numpy array.
+
+    In case of multiple indices corresponding to the maximum value, the
+    result is chosen randomly among those. The random number generator
+    can be seeded by forwarding a seed value: see function 'hash_dist'.
+
+    Parameters
+    ----------
+    a : array_like
+        The source array.
+    random_seed : int
+        RNG seed.
+
+    Returns
+    -------
+    index : int
+        Index of the maximum value.
+
+    Raises
+    ------
+    ValueError : shape mismatch
+        If 'a' has got more than 2 dimensions.
+
+    Notes
+    -----
+    2-D arrays are also supported to avoid multiple RNG initialisation.
+    An array of indices corresponding to the maximum value of each row
+    is then returned.
+    """
+    if a.ndim > 2:
+        raise ValueError("argmaxrnd only accepts arrays of up to 2 dim")
+
+    random = (np.random if random_seed is None
+              else np.random.RandomState(random_seed))
+
+    def f(x): return random.choice((x == np.nanmax(x)).nonzero()[0])
```

Comment:
Please use some vspace here.

You can `import bottlechest as bn; bn.nanmax()` for 50% faster.